### PR TITLE
SALTO-3958: scripted fields

### PR DESCRIPTION
**Pull request by IdoZakk**
Added support for fetch and deployment for scripted fields

---

Added some filters and repaired deployment for all script runner items.
Also Added initial fetch for api_config, commented out.
I have split  the PR to two commits, as one was a folder restructure

---
_Release Notes_: 
Jira Adapter:
* Added support for scripted fields in ScriptRunner

---
_User Notifications_: 
Jira Adapter:
* Added support for scripted fields in ScriptRunner

### CODE REVIEW

**Comment by alonstern** [comment]
There are a couple of values I see in the UI and don't see in the NaCl: Search Term and Issue Type/s. Any idea why we don't get them?

File: packages/jira-adapter/src/filters/script_runner/script_runner_filter.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    // update audit info
+    const currentUserInfo = await getCurrentUserInfo(client)
+
+    changes
+      .filter(isAdditionChange)
+      .filter(isInstanceChange)
+      .map(getChangeData)
+      .filter(instance => SCRIPT_RUNNER_TYPES.includes(instance.elemID.typeName))
+      .forEach(instance => {
+        instance.value.auditData = {
```
Comment (alonstern) [inline packages/jira-adapter/src/filters/script_runner/script_runner_filter.ts:60]: don't we need to delete these values onDeploy?

File: packages/jira-adapter/src/filters/script_runner/workflow/walk_on_scripts.ts [hunk 2]
```diff
@@ -81,4 +83,11 @@ export const walkOnScripts = (
         }
       })
     })
+  instances
```
Comment (alonstern) [inline packages/jira-adapter/src/filters/script_runner/workflow/walk_on_scripts.ts:86]: Is this only relevant to workflows?

File: packages/jira-adapter/src/reference_mapping.ts [hunk 3]
```diff
@@ -881,6 +886,25 @@ export const referencesRules: JiraFieldReferenceDefinition[] = [
...
   },
+  // ScriptRunner
+  {
+    src: { field: 'projects', parentTypes: [SCRIPT_RUNNER_LISTENER_TYPE] },
+    jiraSerializationStrategy: 'key',
+    jiraMissingRefStrategy: 'typeAndValue',
+    target: { type: PROJECT_TYPE },
+  },
+  {
+    src: { field: 'projectKeys', parentTypes: [SCRIPTED_FIELD_TYPE] },
```
Comment (alonstern) [inline packages/jira-adapter/src/reference_mapping.ts:897]: This is a problem that a single scriptedField references multiple projects. We might need to do what we did with automation and project to allow deploying a scriptedField with unresolved references to projects. This is probably a separate ticket/PR (though a blocker IMO)

**Review by alonstern — COMMENTED** [review]
🔥

**Comment by IdoZakk** [comment]
> There are a couple of values I see in the UI and don't see in the NaCl: Search Term and Issue Type/s. Any idea why we don't get them?

If you add a search term it will be added to the NaCl.
The issue type I delete, there is a redundancy between issue type and issue type id

File: packages/jira-adapter/src/filters/script_runner/workflow/walk_on_scripts.ts (hunk 2 above)
Comment (IdoZakk) [reply packages/jira-adapter/src/filters/script_runner/workflow/walk_on_scripts.ts:86]: The first part is for workflows, the second part is for script runner items

File: packages/jira-adapter/src/reference_mapping.ts (hunk 3 above)
Comment (IdoZakk) [reply packages/jira-adapter/src/reference_mapping.ts:897]: It holds also for issue types, I agree we should handle but do not think it is a blocker. I will ask in `jira-dev`

**Comment by alonstern** [comment]
> > There are a couple of values I see in the UI and don't see in the NaCl: Search Term and Issue Type/s. Any idea why we don't get them?
> 
> If you add a search term it will be added to the NaCl. The issue type I delete, there is a redundancy between issue type and issue type id

I do not see either issue type or issue type id in the `Sum_Story_Points_Below_Epic@s` scripted field
```
jira.ScriptedField Sum_Story_Points_Below_Epic@s {
  projectKeys = [
    ...
  ]
  name = "Sum Story Points Below Epic"
  scriptedFieldType = "NUMBER_FIELD"
  codeToRun = '''
/*
* Note: This field should be configured to use a 'Number' return type.
* Get the story points for all story issues in the Epic and display a sum of them on the parent Epic issue.
* This example shows how you can get all of the story issues in the Epic and display a sum of them on the parent Epic issue.
*/

// Check if the issue is an Epic issue
    if (issue.fields.issuetype.name == "Epic") {
        
    // Get the field ids
        def fields = get('/rest/api/2/field')
            .asObject(List)
            .body as List<Map>
                        
            
    // Get the Story Points custom field to use in the script
        def storyPointsField = fields.find { it.name == "Story Points" }?.id
        
    // Handle if the Story Points Field does not exist
        if (storyPointsField == null) {
            logger.info("Story Points field does not exist");
                return;
            }
            
    // Get all issues below the the Epic Issue
        def allStories = get("/rest/agile/1.0/epic/\${issue.key}/issue")
    // The JQL query to return all stories, modify this if you wish to return other issue types inside of the epic as well.
            .queryString("jql", "parentEpic =\${issue.key} and issuetype = 'Story'")
            .queryString("fields", "parent,$storyPointsField")
            .asObject(Map)
            .body
            .issues as List<Map>
            
    // Sum the Story Points for all the Story issues returned
        def estimate = allStories.collect { Map story ->
            story.fields[storyPointsField] ?: 0
        }.sum()
        
    // return the estimate value if it is not null and return 0 if it has no value
        return estimate ?: 0;
    }
'''
  enabled = false
  itemLocation = "atl.jira.view.issue.right.context"
}
```

File: packages/jira-adapter/src/filters/script_runner/workflow/walk_on_scripts.ts (hunk 2 above)
Comment (alonstern) [reply packages/jira-adapter/src/filters/script_runner/workflow/walk_on_scripts.ts:86]: Can we move it out of the workflow folder then?

File: packages/jira-adapter/src/filters/script_runner/script_runner_filter.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          createdByAccountId: currentUserInfo?.userId ?? '',
+          createdTimestamp: getTimeNowAsSeconds(),
```
Comment (alonstern) [inline packages/jira-adapter/src/filters/script_runner/script_runner_filter.ts:62]: Does `getTimeNowAsSeconds` returns the time in UTC or GMT? (and is that what the API expects)

**Comment by IdoZakk** [comment]
> > > There are a couple of values I see in the UI and don't see in the NaCl: Search Term and Issue Type/s. Any idea why we don't get them?
> > 
> > 
> > If you add a search term it will be added to the NaCl. The issue type I delete, there is a redundancy between issue type and issue type id
> 
> I do not see either issue type or issue type id in the `Sum_Story_Points_Below_Epic@s` scripted field
> 
> ```
> jira.ScriptedField Sum_Story_Points_Below_Epic@s {
>   projectKeys = [
>     ...
>   ]
>   name = "Sum Story Points Below Epic"
>   scriptedFieldType = "NUMBER_FIELD"
>   codeToRun = '''
> /*
> * Note: This field should be configured to use a 'Number' return type.
> * Get the story points for all story issues in the Epic and display a sum of them on the parent Epic issue.
> * This example shows how you can get all of the story issues in the Epic and display a sum of them on the parent Epic issue.
> */
> 
> // Check if the issue is an Epic issue
>     if (issue.fields.issuetype.name == "Epic") {
>         
>     // Get the field ids
>         def fields = get('/rest/api/2/field')
>             .asObject(List)
>             .body as List<Map>
>                         
>             
>     // Get the Story Points custom field to use in the script
>         def storyPointsField = fields.find { it.name == "Story Points" }?.id
>         
>     // Handle if the Story Points Field does not exist
>         if (storyPointsField == null) {
>             logger.info("Story Points field does not exist");
>                 return;
>             }
>             
>     // Get all issues below the the Epic Issue
>         def allStories = get("/rest/agile/1.0/epic/\${issue.key}/issue")
>     // The JQL query to return all stories, modify this if you wish to return other issue types inside of the epic as well.
>             .queryString("jql", "parentEpic =\${issue.key} and issuetype = 'Story'")
>             .queryString("fields", "parent,$storyPointsField")
>             .asObject(Map)
>             .body
>             .issues as List<Map>
>             
>     // Sum the Story Points for all the Story issues returned
>         def estimate = allStories.collect { Map story ->
>             story.fields[storyPointsField] ?: 0
>         }.sum()
>         
>     // return the estimate value if it is not null and return 0 if it has no value
>         return estimate ?: 0;
>     }
> '''
>   enabled = false
>   itemLocation = "atl.jira.view.issue.right.context"
> }
> ```

This is a very weird behavior. I will ask them. Seems like the IDs field is not reliable on Ori-salto-test. I added one myself with two issue types and got in the ids a list of more than 10

File: packages/jira-adapter/src/filters/script_runner/script_runner_filter.ts (hunk 4 above)
Comment (IdoZakk) [reply packages/jira-adapter/src/filters/script_runner/script_runner_filter.ts:62]: Yes. This is `EPOCH` so there is no time zone here

File: packages/jira-adapter/src/reference_mapping.ts (hunk 3 above)
Comment (IdoZakk) [reply packages/jira-adapter/src/reference_mapping.ts:897]: I opened a different PR for it, will rebase on it once it is in

**Review by alonstern — APPROVED** [review]
👌 
Are we ok with the noise it is going to make in customers that use script runner?
Also, probably not in the scope of this PR but we should probably the body of the requests on errors and throw a better error to the user (from a quick check it looks like they return `{ error: '<message>' }`), could you open a ticket?
